lucet-concurrency-tests/tests/killswitch.rs

```diff
@@ -0,0 +1,731 @@
+    }
+
+    extern "C" fn do_nothing(_vmctx: *mut lucet_vmctx) -> () {}
+
+    extern "C" fn run_hostcall(vmctx: *mut lucet_vmctx) -> bool {
+        extern "C" {
+            fn real_hostcall(vmctx: *mut lucet_vmctx) -> bool;
+        }
+        unsafe { real_hostcall(vmctx) }
+    }
+
+    extern "C" fn run_yielding_hostcall(vmctx: *mut lucet_vmctx) -> () {
+        extern "C" {
+            fn yielding_hostcall(vmctx: *mut lucet_vmctx) -> ();
+        }
+        unsafe { yielding_hostcall(vmctx) }
+    }
+
+    MockModuleBuilder::new()
+        .with_export_func(MockExportBuilder::new(
+            "onetwothree",
+            FunctionPointer::from_usize(onetwothree as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "infinite_loop",
+            FunctionPointer::from_usize(infinite_loop as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "run_guest",
+            FunctionPointer::from_usize(run_guest as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "do_nothing",
+            FunctionPointer::from_usize(do_nothing as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "run_hostcall",
+            FunctionPointer::from_usize(run_hostcall as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "run_yielding_hostcall",
+            FunctionPointer::from_usize(run_yielding_hostcall as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "fatal",
+            FunctionPointer::from_usize(fatal as usize),
+        ))
+        .with_export_func(MockExportBuilder::new(
+            "hit_sigstack_guard_page",
+            FunctionPointer::from_usize(hit_sigstack_guard_page as usize),
+        ))
+        .build()
+}
+
+/// This test hostcall just needs to exist so that we can call it. `LockTestpoints` provide points
+/// for us to ensure that tests that need to happen inside a hostcall, will happen inside a
+/// hostcall.
+#[lucet_hostcall]
+#[no_mangle]
+pub fn real_hostcall(_vmctx: &mut Vmctx) -> bool {
+    true
+}
+
+/// This test hostcall will immediately yield. This is used to test termination of a
+/// yielded instance.
+#[lucet_hostcall]
+#[no_mangle]
+pub fn yielding_hostcall(vmctx: &mut Vmctx) {
+    vmctx.yield_();
+}
+
+pub fn test_c_with_instrumented_guest_entry<F, R>(dir: &str, cfile: &str, f: F) -> R
+where
+    F: FnOnce(InstanceHandle) -> R,
+{
+    test_ex(|| {
+        unsafe {
+            ENTERING_GUEST = Some(Syncpoint::new());
+        }
+        let module = test_module_c(dir, cfile).expect("build and load module");
+        let region = MmapRegion::create(1, &Limits::default()).expect("region can be created");
+
+        let inst = region
+            .new_instance(module)
+            .expect("instance can be created");
+
+        f(inst)
+    })
+}
+
+pub fn test_exclusive_instance_with_instrumented_guest_entry<F, R>(f: F) -> R
```

Comment:
I am curious about the difference between an **instance** and an **exclusive instance**.

Reply:
this is more a 'test' and 'test exclusive' delineation to control process-wide state - this is taken from `lucet-runtime-tests` like much else, where it's to control what signal handler is installed there. Here, it's also necessary to ensure use of `ENTERING_GUEST` doesn't happen in an unsafe way (with multiple threads clobbering the shared global).

This deserves a note here and on `test_instance_with_...` too :D

Reply:
This should be read as `test_...` and `test_exclusive..`, rather than `exclusive` being a modifier on "instance". I've added some documentation on these helpers to clarify, but just for the viewers at home: some tests require access to process-wide resources (signal handler, the `ENTERING_GUEST` syncpoint), so to guarantee the state and safety of those a test will need a guarantee of exclusive access.

`test_ex` and `test_nonex` in `lucet-runtime-tests` give this same building block, but `lucet-concurrency-tests` adds a new item to manage with `ENTERING_GUEST` and these are convenient points to move out repeated instance setup.